contracts/contracts/token/TokenMigration.sol

```diff
@@ -0,0 +1,110 @@
+pragma solidity ^0.4.23;
+
+import "../../../node_modules/openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "../../../node_modules/openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+import "./OriginToken.sol";
+
+contract TokenMigration is Ownable {
+  OriginToken public fromToken;
+  OriginToken public toToken;
+  mapping (address => bool) migrated;
+  bool public finished;
+
+  event Migrated(address indexed account, uint256 balance);
+  event MigrationFinished();
+  
+  modifier notFinished() {
+    require(!finished, "migration already finished");
+    _;
+  }
+
+  constructor(OriginToken _fromToken, OriginToken _toToken) public {
+    owner = msg.sender;
+    fromToken = _fromToken;
+    toToken = _toToken;
+  }
+
+  function migrateAccounts(address[] _holders) public onlyOwner notFinished {
+    for (uint i = 0; i < _holders.length; i++) {
```

Comment:
Doing this in one for loop from solidity transaction probably isn't going work because of gas limits. If we have 20,000 token holders this will gas limit out far before that point.

I think the usual practice is to store a migrated-so-far index in the contract, and do X amount of migrations each transaction, updating the migrated-so-far index at the end of each mini-loop. Then just call the thing until it's done.

Reply:
My plan is to call `migrateAccounts` for batches of accounts from a JS command-line tool through `web3`. Some JS needs to be involved, because the list of `address`es will come from retrieving past logs, which are only accessible to the JS. The JS would then perform checkpointing to ensure that it can restart upon crash or interruption.

I'm not sure how I could store an index, unless I stored all accounts to migrate as an array in the contract. I'm not clear on what the benefit to doing that is though.

Reply:
I didn't pay enough attention when reading the code. Yeah, I think your way will work.  I like the token totals check being an integrated part of the finalize step.